src/v/kafka/server/quota_translator.h

```diff
@@ -0,0 +1,103 @@
+/// tracker_key may be client id specific even though the matching entity_key is
+/// the default client quota. In this case, we may have multiple independent
+/// rate trackers for each unique client with all of these rate tracking having
+/// the same shared quota limit
+using tracker_key = std::variant<k_client_id, k_group_name>;
+
+/// quota_limits describes the limits applicable to a tracker_key
+struct quota_limits {
+    std::optional<uint64_t> produce_limit;
+    std::optional<uint64_t> fetch_limit;
+    std::optional<uint64_t> partition_mutation_limit;
+};
+
+/// quota_translator is responsible for providing quota_manager with a
+/// simplified interface to the quota configurations
+///  * It is responsible for translating the quota-specific request context (for
+///  now the client.id header field) into the applicable tracker_key.
+///  * It is also responsible for finding the quota limits of a tracker_key
+class quota_translator {
+public:
+    using on_change_fn = std::function<void()>;
+
+    quota_translator();
+
+    /// Returns the quota tracker key used for produce requests
+    /// Note: because the client quotas configured for produce/fetch/pm might be
+    /// different, the tracker_key for produce/fetch/pm might be different
+    tracker_key
+    get_produce_key(const std::optional<std::string_view>& client_id);
+
+    /// Returns the quota tracker key used for fetch requests
+    /// Note: because the client quotas configured for produce/fetch/pm might be
+    /// different, the tracker_key for produce/fetch/pm might be different
+    tracker_key get_fetch_key(const std::optional<std::string_view>& client_id);
+
+    /// Returns the quota tracker key used for partition mutation requests
+    /// Note: because the client quotas configured for produce/fetch/pm might be
+    /// different, the tracker_key for produce/fetch/pm might be different
+    tracker_key get_partition_mutation_key(
+      const std::optional<std::string_view>& client_id);
+
+    /// Finds the limits applicable to the given quota tracker key
+    quota_limits find_quota_value(const tracker_key&);
+
+    /// `watch` can be used to register for quota changes
+    void watch(on_change_fn&& fn) { _on_change = std::move(fn); };
+
+private:
+    using quota_config
+      = std::unordered_map<ss::sstring, config::client_group_quota>;
+
+    int64_t get_client_target_produce_tp_rate(const tracker_key& quota_id);
+    std::optional<int64_t>
+    get_client_target_fetch_tp_rate(const tracker_key& quota_id);
+
+    config::binding<uint32_t> _default_target_produce_tp_rate;
+    config::binding<std::optional<uint32_t>> _default_target_fetch_tp_rate;
+    config::binding<std::optional<uint32_t>> _target_partition_mutation_quota;
+    config::binding<quota_config> _target_produce_tp_rate_per_client_group;
+    config::binding<quota_config> _target_fetch_tp_rate_per_client_group;
+
+    std::optional<on_change_fn> _on_change;
```

Comment:
It might be simpler to do something like the following and then always invoke it.
```suggestion
    on_change_fn _on_change{[](){}};

Reply:
I went with removing the `_on_change` field altogether: https://github.com/redpanda-data/redpanda/pull/18743#discussion_r1622575577